macros/src/typetag.rs

```diff
@@ -0,0 +1,165 @@
+use crate::{Result, util::parse_config_args};
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{DeriveInput, Field, FieldMutability, Ident, Lit, parse_quote, parse2, spanned::Spanned};
+
+const DEFAULT_DISCRIMINATOR: &str = "type";
+
+pub fn type_tag(attr: TokenStream, item: &TokenStream) -> Result<TokenStream, syn::Error> {
+    let mut ast = parse2::<DeriveInput>(item.clone())?;
+
+    let syn::Data::Struct(struct_data) = &mut ast.data else {
+        return Err(syn::Error::new_spanned(
+            &ast,
+            "type tag can only be derived for structs",
+        ));
+    };
+
+    let syn::Fields::Named(fields) = &mut struct_data.fields else {
+        return Err(syn::Error::new_spanned(
+            &ast,
+            "type tag can only be derived for named structs",
+        ));
+    };
+
+    let inputs = parse_config_args(attr)?;
+
+    let Some(literal) = inputs.get("tag") else {
```

Comment:
I find this name confusing because in serde the "tag" is the name of the field, but you use it as the value.

Reply:
[465e085](https://github.com/geo-engine/geoengine/pull/1013/commits/465e0855cc0bdce1f11c0a107588e0b3a510f076)